lib/codeintel/lsif/protocol/reader/lsif_typed.go

```
@@ -0,0 +1,452 @@
+	}
+	if isExported {
+		g.emitEdge("textDocument/definition", Edge{OutV: ids.ResultSet, InV: ids.DefinitionResult})
+	}
+	g.emitEdge("textDocument/references", Edge{OutV: ids.ResultSet, InV: ids.ReferenceResult})
+	g.emitEdge("textDocument/hover", Edge{OutV: ids.ResultSet, InV: ids.HoverResult})
+	if monikerKind == "export" || monikerKind == "import" {
+		g.emitMonikerVertex(info.Symbol, monikerKind, ids.ResultSet)
+	}
+	return ids
+}
+
+// emitDocument emits all range vertices for the `lsif_typed.Occurrence` in the provided document, along with
+// associated `item` edges to link ranges with result sets.
+func (g *graph) emitDocument(index *lsif_typed.Index, doc *lsif_typed.Document) {
+	uri := filepath.Join(index.Metadata.ProjectRoot, doc.RelativePath)
+	documentID := g.emitVertex("document", uri)
+
+	documentSymbolTable := map[string]*lsif_typed.SymbolInformation{}
+	localSymbolInformationTable := map[string]symbolInformationIDs{}
+	for _, info := range doc.Symbols {
+		documentSymbolTable[info.Symbol] = info
+
+		// Build symbol information table for Document-local symbols only.
+		if lsif_typed.IsLocalSymbol(info.Symbol) {
+			localSymbolInformationTable[info.Symbol] = g.emitResultSet(info, "")
+		}
+
+		// Emit "implementation" monikers for external symbols (monikers with kind "import")
+		for _, relationship := range info.Relationships {
+			if relationship.IsImplementation {
+				relationshipIDs := g.getOrInsertSymbolInformationIDs(relationship.Symbol, localSymbolInformationTable)
+				if relationshipIDs.DefinitionResult > 0 {
+					// Not an imported symbol
+					continue
+				}
+				infoIDs := g.getOrInsertSymbolInformationIDs(info.Symbol, localSymbolInformationTable)
+				g.emitMonikerVertex(relationship.Symbol, "implementation", infoIDs.ResultSet)
+			}
+		}
+	}
+
+	var rangeIDs []int
+	for _, occ := range doc.Occurrences {
+		rangeID, err := g.emitRange(occ.Range)
+		if err != nil {
+			// Silently skip invalid ranges.
```

Can we accumulate the errors and log them in a single line later as a warning? IIUC silently ignoring errors here would make it harder to diagnose issues where an indexer emits some incorrect ranges.

---

Alternately, even if this isn't the default, it would be nice to log this in a `--verbose` mode (doesn't have to be part of this PR).

---

I added a TODO linking to this issue here https://github.com/sourcegraph/sourcegraph/issues/31415